Add Task Tag Concurrency Limiting Doc (#1596)
PrefectHQ/prefect: docs/cloud/concepts/concurrency-limiting.md

# Task Concurrency Limiting

Oftentimes there are situations in which users want to actively prevent too many tasks from running simultaneously; for example, if many tasks across multiple Flows are designed to interact with a database that only allows 10 max connections, we want to ensure that no more than 10 tasks which connect to this database are running at any given time.

> Prefect Cloud has built-in functionality for doing this; Tasks can be "tagged" with as many tags as you wish, and each tag can optionally be provided a concurrency limit.  _All_ task tag concurrency limits will be respected at runtime.

Author: ```suggestion
Prefect Cloud has built-in functionality for achieving this; Tasks can be "tagged" with as many tags as you wish, and each tag can optionally be provided a concurrency limit.  _All_ task tag concurrency limits will be respected at runtime.
```

> Prefect Cloud has built-in functionality for achieving this; Tasks can be "tagged" with as many tags as you wish, and each tag can optionally be provided a concurrency limit.  _All_ task tag concurrency limits will be respected at runtime.

Reviewer: Minor clarification by what "All" means -- Perhaps 
```suggestion
Prefect Cloud has built-in functionality for achieving this; Tasks can be "tagged" with as many tags as you wish, and each tag can optionally be provided a concurrency limit. If a task has multiple tags, it will only run if _all_ tags have available concurrency.
```

> 

Reviewer: Also, a suggestion to clarify behavior:

> Tag concurrency limits are enforced globally across your entire team, and tags without explicit limits are considered to have unlimited concurrency.

:::

To remove all concurrency limits on a tag, simply issue:

```graphql
mutation {
  deleteTaskTagLimit(input: { limitId: "uuid-returned-from-above" }) {
    success
  }
}
```

## Querying Concurrency Limits

If you wish to query for the currently set limit on a tag, or see _all_ of your limits across all of your tags, you can similarly use both Prefect's Python Client as well as GraphQL directly.

### Python Client

```python
from prefect import Client

client = Client()

# retrieve the current limit on the "database" tag
# a return value of `None` means no limit is currently set
client.get_task_tag_limit("database")
```

### GraphQL <Badge text="GQL"/>

GraphQL allows you to retrieve more, including:
- _all_ of your tag limits
- your tag limit IDs (useful for deleting limits)

```graphql
query {
    task_tag_limits(where: {tag: {_eq: "webservice" } }){
        limit
        id
    }
}
```

::: tip In GraphQL, everything is equal to `null`
To retrieve _all_ limits across all tags, simply replace the value of `"webservice"` above with `null`.
:::

## Execution Behavior

> Task tag limits are checked whenever a Task Run for a tagged task requests to enter a ["Running" state](../../core/concepts/states.html) with Prefect Cloud. If there are no concurrency slots available for any one of your Task's tags, the Task will instead enter a "Queued" state.  The process which is currently running your Task will then attempt to re-enter a Running state every 30 seconds (this value is configurable via `config.cloud.queue_interval` in [Prefect Configuration](../../core/concepts/configuration.html)).  Additionally, if this process ever fails, Prefect Cloud will recreate a runner every 10 minutes, which will then attempt to rerun your task on the specified queue interval, all until a concurrency slot becomes available.

Reviewer: ```suggestion
Task tag limits are checked whenever a task run attempts to enter a [`Running` state](../../core/concepts/states.html) in Prefect Cloud. If there are no concurrency slots available for any one of your Task's tags, the Task will instead enter a `Queued` state.  The same Python process that is attempting running your Task will then attempt to re-enter a `Running` state every 30 seconds (this value is configurable via `config.cloud.queue_interval` in [Prefect Configuration](../../core/concepts/configuration.html)).  Additionally, if that process ever fails, Prefect Cloud will create a new runner every 10 minutes, which will then attempt to rerun your task on the specified queue interval. This process will repeat until all requested concurrency slots become available.
```
